File: tests/sample_project_desktop/client_example.py
"""Python client example for the matrx-orm API server.

Shows how to authenticate and make RPC calls from Python.
This pattern works for:
  - Python desktop apps (e.g. with PySide/PyQt)
  - Testing and scripting
  - Any Python service that needs to talk to the API
"""

from __future__ import annotations

import asyncio
import json
from typing import Any

import aiohttp


class MatrxClient:
    """Simple async client for the matrx-orm API server.

    Usage::

        async with MatrxClient("http://127.0.0.1:8745", secret="...") as client:
            users = await client.call("users.load_items", status="active")
            user = await client.call("users.create_item",
                username="alice",
                email="alice@example.com",
            )
    """

    def __init__(self, base_url: str, secret: str) -> None:
        self._base_url = base_url.rstrip("/")
        self._secret = secret
        self._token: str | None = None
        self._session: aiohttp.ClientSession | None = None
        self._request_id = 0

    async def __aenter__(self) -> MatrxClient:
        self._session = aiohttp.ClientSession()
        await self._authenticate()
        return self

    async def __aexit__(self, *args: Any) -> None:
        if self._session:
            await self._session.close()

    async def _authenticate(self) -> None:
        """Exchange the shared secret for a bearer token."""
        assert self._session is not None
        async with self._session.post(
            f"{self._base_url}/auth/token",
            json={"secret": self._secret, "client_id": "python-client"},
        ) as resp:
            if resp.status != 200:
                raise RuntimeError(f"Authentication failed: {resp.status}")
            data = await resp.json()
            self._token = data["token"]

    async def call(self, method: str, **params: Any) -> Any:
        """Make an RPC call and return the result.

        Raises RuntimeError on RPC errors.
        """
        assert self._session is not None
        self._request_id += 1

        payload = {
            "method": method,
            "params": params,
            "id": self._request_id,
        }

        async with self._session.post(
            f"{self._base_url}/rpc",
            json=payload,
            headers={"Authorization": f"Bearer {self._token}"},
        ) as resp:
            text = await resp.text()
            data = json.loads(text)

            if data.get("error"):
                err = data["error"]
                raise RuntimeError(
                    f"RPC Error {err['code']}: {err['message']}"
                )

            return data.get("result")

    async def call_batch(self, calls: list[tuple[str, dict]]) -> list[Any]:
        """Make multiple RPC calls in a single HTTP request.

        Parameters
        ----------
        calls : list[tuple[str, dict]]
            List of (method, params) tuples.

        Returns
        -------
        list[Any]
            List of results in the same order as the input.
        """
        assert self._session is not None

        batch = []
        for method, params in calls:
            self._request_id += 1
            batch.append({
                "method": method,
                "params": params,
                "id": self._request_id,
            })

        async with self._session.post(
            f"{self._base_url}/rpc",
            json=batch,
            headers={"Authorization": f"Bearer {self._token}"},
        ) as resp:
            text = await resp.text()
            responses = json.loads(text)
            return [r.get("result") for r in responses]

    async def health(self) -> dict:
        """Check server health (no auth required)."""
        assert self._session is not None
        async with self._session.get(f"{self._base_url}/health") as resp:
            return await resp.json()


# ── Example usage ──────────────────────────────────────────────────────────

async def main() -> None:
    # The secret is printed by the server on startup (or set via API_SECRET env var)
    SECRET = "change-me-to-a-real-secret"
    SERVER_URL = "http://127.0.0.1:8745"

    async with MatrxClient(SERVER_URL, secret=SECRET) as client:
        # Health check
        health = await client.health()
        print(f"Server health: {health}")

        # Create a user
        user = await client.call(
            "users.create_item",
            username="alice",
            email="alice@example.com",
        )
        print(f"Created user: {user}")

        # Load all users
        users = await client.call("users.load_items")
        print(f"All users: {users}")

        # Create a note
        note = await client.call(
            "notes.create_item",
            title="My First Note",
            content="Hello from the desktop app!",
            user_id=user["id"],
        )
        print(f"Created note: {note}")

        # Load note with author (FK prefetch via ModelView)
        note_with_author = await client.call(
            "notes.load_item",
            id=note["id"],
        )
        print(f"Note with author: {note_with_author}")

        # Batch request — multiple calls in one HTTP round-trip
        results = await client.call_batch([
            ("users.count", {}),
            ("notes.count", {}),
            ("app.info", {}),
        ])
        print(f"User count: {results[0]}")
        print(f"Note count: {results[1]}")
        print(f"App info: {results[2]}")

        # Update
        updated = await client.call(
            "notes.update_item",
            item_id=note["id"],
            title="Updated Title",
        )
        print(f"Updated note: {updated}")

        # Delete
        deleted = await client.call(
            "notes.delete_item",
            item_id=note["id"],
        )
        print(f"Deleted: {deleted}")


if __name__ == "__main__":
    asyncio.run(main())

File: tests/sample_project_desktop/client_example.ts
/**
 * TypeScript client for the matrx-orm API server.
 *
 * Copy this file into your Electron / Tauri / web frontend project.
 * No external dependencies — uses the Fetch API and native WebSocket.
 *
 * Usage:
 *   const client = new MatrxClient("http://127.0.0.1:8745", "your-secret");
 *   await client.authenticate();
 *
 *   const users = await client.call("users.load_items", { status: "active" });
 *   const user = await client.call("users.create_item", {
 *     username: "alice",
 *     email: "alice@example.com",
 *   });
 */

// ── Types ─────────────────────────────────────────────────────────────────

interface RPCRequest {
  method: string;
  params: Record<string, unknown>;
  id: number;
}

interface RPCError {
  code: number;
  message: string;
  data?: Record<string, unknown>;
}

interface RPCResponse<T = unknown> {
  result: T | null;
  error: RPCError | null;
  id: number | null;
}

interface AuthResponse {
  token: string;
  expires_in: number;
}

interface HealthResponse {
  status: string;
  managers: string[];
}

// ── Error codes ───────────────────────────────────────────────────────────

export enum ErrorCode {
  PARSE_ERROR = -32700,
  INVALID_REQUEST = -32600,
  METHOD_NOT_FOUND = -32601,
  INVALID_PARAMS = -32602,
  INTERNAL_ERROR = -32603,
  NOT_FOUND = 1001,
  VALIDATION_ERROR = 1002,
  INTEGRITY_ERROR = 1003,
  AUTH_REQUIRED = 1004,
  AUTH_INVALID = 1005,
  PERMISSION_DENIED = 1006,
  RATE_LIMITED = 1007,
  CONNECTION_ERROR = 1008,
  QUERY_ERROR = 1009,
}

// ── Client error ──────────────────────────────────────────────────────────

export class MatrxRPCError extends Error {
  code: number;
  data?: Record<string, unknown>;

  constructor(error: RPCError) {
    super(error.message);
    this.name = "MatrxRPCError";
    this.code = error.code;
    this.data = error.data;
  }
}

// ── HTTP Client ───────────────────────────────────────────────────────────

export class MatrxClient {
  private baseUrl: string;
  private secret: string;
  private token: string | null = null;
  private tokenExpiry: number = 0;
  private requestId = 0;

  constructor(baseUrl: string, secret: string) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.secret = secret;
  }

  /**
   * Exchange the shared secret for a bearer token.
   * Call this before making any RPC calls.
   */
  async authenticate(clientId = "desktop"): Promise<void> {
    const resp = await fetch(`${this.baseUrl}/auth/token`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        secret: this.secret,
        client_id: clientId,
      }),
    });

    if (!resp.ok) {
      throw new Error(`Authentication failed: ${resp.status}`);
    }

    const data: AuthResponse = await resp.json();
    this.token = data.token;
    this.tokenExpiry = Date.now() + data.expires_in * 1000;
  }

  /**
   * Check if the current token is still valid.
   */
  get isAuthenticated(): boolean {
    return this.token !== null && Date.now() < this.tokenExpiry;
  }

  /**
   * Re-authenticate if the token has expired.
   */
  private async ensureAuth(): Promise<void> {
    if (!this.isAuthenticated) {
      await this.authenticate();
    }
  }

  /**
   * Make a single RPC call.
   *
   * @param method - The RPC method (e.g. "users.load_item")
   * @param params - Keyword arguments as an object
   * @returns The result from the server
   * @throws MatrxRPCError on RPC errors
   */
  async call<T = unknown>(
    method: string,
    params: Record<string, unknown> = {}
  ): Promise<T> {
    await this.ensureAuth();
    this.requestId++;

    const payload: RPCRequest = {
      method,
      params,
      id: this.requestId,
    };

    const resp = await fetch(`${this.baseUrl}/rpc`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.token}`,
      },
      body: JSON.stringify(payload),
    });

    const data: RPCResponse<T> = JSON.parse(await resp.text());

    if (data.error) {
      throw new MatrxRPCError(data.error);
    }

    return data.result as T;
  }

  /**
   * Make multiple RPC calls in a single HTTP request.
   *
   * @param calls - Array of [method, params] tuples
   * @returns Array of results in the same order
   */
  async callBatch(
    calls: Array<[string, Record<string, unknown>]>
  ): Promise<unknown[]> {
    await this.ensureAuth();

    const batch: RPCRequest[] = calls.map(([method, params]) => {
      this.requestId++;
      return { method, params, id: this.requestId };
    });

    const resp = await fetch(`${this.baseUrl}/rpc`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.token}`,
      },
      body: JSON.stringify(batch),
    });

    const responses: RPCResponse[] = JSON.parse(await resp.text());
    return responses.map((r) => {
      if (r.error) throw new MatrxRPCError(r.error);
      return r.result;
    });
  }

  /**
   * Health check (no auth required).
   */
  async health(): Promise<HealthResponse> {
    const resp = await fetch(`${this.baseUrl}/health`);
    return resp.json();
  }
}

// ── WebSocket Client ──────────────────────────────────────────────────────

type WSMessageHandler = (response: RPCResponse) => void;

export class MatrxWSClient {
  private baseUrl: string;
  private token: string;
  private ws: WebSocket | null = null;
  private requestId = 0;
  private pending = new Map<number, {
    resolve: (value: unknown) => void;
    reject: (error: Error) => void;
  }>();
  private onMessage?: WSMessageHandler;

  constructor(baseUrl: string, token: string) {
    // Convert http:// to ws://
    this.baseUrl = baseUrl.replace(/^http/, "ws");
    this.token = token;
  }

  /**
   * Connect and authenticate via WebSocket.
   */
  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.ws = new WebSocket(`${this.baseUrl}/ws`);

      this.ws.onopen = () => {
        // Send auth token as first message
        this.ws!.send(JSON.stringify({ token: this.token }));
      };

      this.ws.onmessage = (event) => {
        const data = JSON.parse(event.data);

        // Auth response
        if (data.type === "auth_ok") {
          resolve();
          return;
        }
        if (data.type === "auth_error") {
          reject(new Error(data.error));
          return;
        }

        // RPC response
        const response = data as RPCResponse;
        const id = response.id as number;
        const handler = this.pending.get(id);
        if (handler) {
          this.pending.delete(id);
          if (response.error) {
            handler.reject(new MatrxRPCError(response.error));
          } else {
            handler.resolve(response.result);
          }
        }

        // General message handler
        this.onMessage?.(response);
      };

      this.ws.onerror = () => reject(new Error("WebSocket connection failed"));
      this.ws.onclose = () => {
        // Reject all pending requests
        for (const [, handler] of this.pending) {
          handler.reject(new Error("WebSocket closed"));
        }
        this.pending.clear();
      };
    });
  }

  /**
   * Make an RPC call over the WebSocket.
   */
  async call<T = unknown>(
    method: string,
    params: Record<string, unknown> = {}
  ): Promise<T> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error("WebSocket is not connected");
    }

    this.requestId++;
    const id = this.requestId;

    const payload: RPCRequest = { method, params, id };

    return new Promise((resolve, reject) => {
      this.pending.set(id, {
        resolve: resolve as (value: unknown) => void,
        reject,
      });
      this.ws!.send(JSON.stringify(payload));
    });
  }

  /**
   * Close the WebSocket connection.
   */
  close(): void {
    this.ws?.close();
    this.ws = null;
  }
}

// ── Example usage ─────────────────────────────────────────────────────────

/*
// In your Electron renderer or Tauri frontend:

const client = new MatrxClient("http://127.0.0.1:8745", "your-secret");
await client.authenticate();

// CRUD operations
const user = await client.call<{ id: string; username: string }>(
  "users.create_item",
  { username: "alice", email: "alice@example.com" }
);

const users = await client.call<Array<{ id: string; username: string }>>(
  "users.load_items",
  { is_active: true }
);

// Batch operations
const [userCount, noteCount] = await client.callBatch([
  ["users.count", {}],
  ["notes.count", {}],
]);

// WebSocket for real-time updates
const wsClient = new MatrxWSClient("http://127.0.0.1:8745", client.token);
await wsClient.connect();
const result = await wsClient.call("users.load_items");
wsClient.close();
*/
